### ISSUE TITLE
Audio playback feature

### ISSUE THREAD

**Issue by SevenBlocks** [opensheetmusicdisplay/opensheetmusicdisplay#258]
Are there plans to add playback support--something that moves the cursor, plays the current notes, and waits until the next note should be played? Or is that considered out of scope for this project? I'm interested in potentially working on this feature.

**Comment by matt-uib** [comment]
Yes this is on our plan. Actually we will very soon start working on a plugin infrastructure for osmd. Our plan here is, that we can also offer services and plugins for a license fee (see also https://opensheetmusicdisplay.org/develop/roadmap/ ) and also other people can write free or paid plugins. The first plugin on our list is a playback engine which we will tackle this summer in July/August.
Planned Playback Features:
  
-    Set start position (via click onto a note on the score)
-    set end position (for looping)
-    set speed in BPM
-    play back a metronome together with the music (with pre-count)
-    Playback of given chord symbols
-    Solo, Mute switches and volume settings for all instruments, the metronome and the chords (if avaliable)
-    React on repetitions (backjumps), on dynamics (p, ff) and tempo instructions

@SevenBlocks, It is very likely that we need help on this feature, but this depends on funding and some other parameters. If you are generally interested to join in for this, please leave me here your email so that we can contact you on this in early July.

**Comment by infojunkie** [comment]
Hello, I'm also interested in a playback plugin. I've written similar code for VexFlow here: https://ethereum.karimratib.me:8080/ - it supports local audio playback + Web MIDI output.

**Comment by sschmidTU** [comment]
I'm probably going to work on Playback soon, provided that the plugin infrastructure will be ready.
@infojunkie your VexFlow player looks nice, great work! Is it open source? i'd love to take a look at it.
@SevenBlocks @acidjunk Great to hear you want to contribute! We'll keep you updated. (acid, i assumed you wanted to contribute too.)

**Comment by svenn71** [comment]
I think www.soundslice.com is a nice reference when it comes to look+feel+sound of displaying musicxml in a browser.

**Comment by bneumann** [comment]
@svenn71 Jupp that's a cool website but they are a company and this is an open source project in the very beginning. So while it is cool to take it as a reference I find it a bit over ambitious to create the same look and feel with a handful part time volunteers. But feel free to contribute to get us there 😉

**Comment by ifredom** [comment]
> * Set start position (via click onto a note on the score)
> * set end position (for looping)


hi。Where can I find this api?

**Comment by sschmidTU** [comment]
It's a planned feature, not yet available.

**Comment by sschmidTU** [comment]
The OSMD team is working on Audio playback as a dedicated project for 2020, more info coming soon.

**Comment by tcdaly** [comment]
I have a project for which the audio playback feature would be very useful.  Looking forward to it!

**Comment by fredmcgroarty** [comment]
@sschmidTU I'd be really happy to contribute to this project too.

**Comment by conde2** [comment]
I'm so excited for this feature, is this still a thing for this year ? Can't wait to see this !!

Just a suggestion, would be very useful to have events that we can connect whenever a note is going to be played, and it returns the Note object with all the information.

Also thanks all contributors for this awesome project.

**Comment by sschmidTU** [comment]
Yes, we are hard at work on this, hoping for a release rather soon, this year!
Suggestion noted, would indeed be useful.

**Comment by alex-sherwin** [comment]
I started learning Piano Dec 2020 (first instrument ever for me), learning to read sheet music, etc.

As a software developer I was appalled at the status quo of Sheet Music (particularly the marketplaces for them), that intentionally make it difficult to have structured data to do something interesting with it with software.

I like apps like FlowKey & Simply Piano with their "wait for user" features listening for MIDI signals, but they are too restrictive with their small'ish catalog of custom-made sheet music for their platform.  Not to take away from their apps, they are great, and the video overlay with finger information is awesome for beginners, but I wanted to be able to have a similar feature set for *any* music.

Long story short, after some searching I figured the best combination of software was WebMIDI + MusicXML + OpenSheetMusicDisplay.  

I came across this issue and found that it's been being "worked on" for quite some time, without much info on what that actual feature might even contain, so I went ahead and created it myself.

So I have a complete working React (100% TypeScript) + OpenSheetMusicDisplay + MusicXML + WebMIDI app that has:

* "Wait for input" (i.e. waits at each position for MIDI signals for the notes to be played by the user)
* Looping between arbitrary positions (i.e. learn a subset of measures at a time)
* Playback over MIDI (plays back the sheet music via your MIDI device)
* On-screen 88-key piano that shows what keys are waiting to be pressed, and what keys you are actively pressing
* Works on iOS/mobile Safari (by using https://github.com/mizuhiki/WebMIDIAPIShimForiOS as a WebMIDI shim)
* CSS styling tailored to work on either iPad or iPhone

It's actually probably good enough that I could polish it up and sell it on the App Store, but my intent was really just to make something for myself.

To get it to that state it would basically only need some native iOS file management features to make it easy to add new MusicXML files. Currently, I'm just hard-coding my own MusicXML files in my webpack bundle output.

Anyways, I'd be willing to contribute/share the complete app if it would be useful to this project to utilize
